### ISSUE TITLE
Possible infinite loop while typing in the editor

### ISSUE THREAD

**Issue by sam974** [eclipse/xtext-eclipse#678]
With my language definition, I made my product frozen while typing in the editor.

https://github.com/eclipse/xtext-eclipse/blob/89f1c0179d78f2c2f28b1cb8ea10916ec35ea55a/org.eclipse.xtext.ui/src/org/eclipse/xtext/ui/editor/model/DocumentTokenSource.java#L291

The infinite loop can occur when the assigned token is an invalid one which was not consumed.
https://github.com/eclipse/xtext-eclipse/blob/89f1c0179d78f2c2f28b1cb8ea10916ec35ea55a/org.eclipse.xtext.ui/src/org/eclipse/xtext/ui/editor/model/DocumentTokenSource.java#L311

This state occurs in my case when an org.antlr.runtime.EarlyExitException is raised which is [not recovered.](https://github.com/eclipse/xtext-core/blob/dda480aab67be6f99b407ed6c19ba9b1c125f0c8/org.eclipse.xtext/src/org/eclipse/xtext/parser/antlr/Lexer.java#L75)

**Comment by kthoms** [comment]
How to reproduce this? Could you create a minimal example and tell where and how to edit it to fall into this loop? Otherwise it would be unlikely that we can fix it until we ourselves run into this issue (which we obviously did not so far ;) )

**Comment by miklossy** [comment]
I also encountered a lexer infinitie loop problem recently, see [GEF DOT Bugreport](https://bugs.eclipse.org/bugs/show_bug.cgi?id=529703) and [Xtext forum thread](https://www.eclipse.org/forums/index.php/t/958598/).

@sam974: You could really help us by providing a small example to be able to reproduce your issue. The [Eclipse GEF DOT](http://www.eclipse.org/gef/) open source project already provides some test cases to verify that none of the lexers go into infinite loops. Feel free to get some inspiration from them:

- Runtime lexer test cases: [DotLexerTests.xtend](http://github.com/eclipse/gef/blob/master/org.eclipse.gef.dot.tests/src/org/eclipse/gef/dot/tests/DotLexerTests.xtend),  [DotHtmlLabelLexerTests.xtend](http://github.com/eclipse/gef/blob/master/org.eclipse.gef.dot.tests/src/org/eclipse/gef/dot/tests/DotHtmlLabelLexerTests.xtend)
	
- Content assistant lexer test cases: [DotHtmlLabelContentAssistLexerTests.xtend](http://github.com/eclipse/gef/blob/master/org.eclipse.gef.dot.tests/src/org/eclipse/gef/dot/tests/DotHtmlLabelContentAssistLexerTests.xtend)
	
- Highlighting lexer test cases: [DotHtmlLabelHighlightingLexerTests.xtend](https://github.com/eclipse/gef/blob/master/org.eclipse.gef.dot.tests/src/org/eclipse/gef/dot/tests/DotHtmlLabelHighlightingLexerTests.xtend)
	
@cdietrich, @kthoms: I am wondering why the [fix](http://github.com/eclipse/xtext-core/commit/34d61d4d80c6992045ed013a0ac4eb337a1f4e87) has been applied only on the runtime lexer and why it is not applied on the content assistant/highlighting lexers.

**Comment by szarnekow** [comment]
Good catch. We should certainly add a test and apply the fix to all lecker base types

**Comment by sam974** [comment]
To reproduce: 
- just create an Xtext project with the attached grammar
[Api.xtext.txt](https://github.com/eclipse/xtext-eclipse/files/2017553/Api.xtext.txt)

- once your runtime launched, create a 'foo.api' file, open it  and then start to insert in the editor  the keyword USETYPES followed immediately by the dollar character;  the infinite loop should occur...

**Comment by miklossy** [comment]
Based on the description above, I could reproduce the bug and made the editor frozen. 

I created a small [Xtext project](https://github.com/itemis/xtext-issues/tree/issues/xtext-eclipse/678) with corresponding JUnit test cases. Some of them pass, some of them fails. @sam974  Feel free to fork the repository and provide further (both passing or failing) test cases. The more the test cases, the better.

**Comment by sam974** [comment]
The Xtext project has been forked and a [commit](https://github.com/itemis/xtext-issues/pull/1/commits) with 2 test cases has been pushed

**Comment by sam974** [comment]
As a [workaround ](https://github.com/sam974/xtext-core/commit/ac43d59a4a0a899fc6a352bf92239249e5dfdfc8)to prevent infinite loop, recover is called on EarlyExitException.

**Comment by miklossy** [comment]
I uploaded another project that represents an infinite loop, see https://github.com/eclipse/xtext-eclipse/issues/690

**Step to reproduce:**
1. create an empty test.xml file
2. open it with the Xml Editor (that is based on Xtext framework)
3. enter `<a b="" />`
4. move the cursor between the quotes and invoke the content assistant

that makes Eclipse to freeze.

That use case has been captured in the [XmlContentAssistTests](https://github.com/itemis/xtext-issues/blob/ec279a3da1d9cf0a5d392879c54b7e5e6dfb621e/org.eclipse.xtext.example.xmleditor.ui.tests/src/org/eclipse/xtext/example/xmleditor/ui/tests/XmlContentAssistTests.xtend#L18-L20) and [XmlContentAssistLexerTests](https://github.com/itemis/xtext-issues/blob/issues/xtext-eclipse/690/org.eclipse.xtext.example.xmleditor.ui.tests/src/org/eclipse/xtext/example/xmleditor/ui/tests/XmlContentAssistLexerTests.xtend) test cases.

**Observation:** The stateful lexer falls into an infinite loop due to an unhandled `org.antlr.runtime.FailedPredicateException`.

**Solution:** The [fix](https://github.com/eclipse/xtext-core/commit/34d61d4d80c6992045ed013a0ac4eb337a1f4e87) has been already applied on the runtime lexer and highlighting lexer, but not yet on the content assistant lexer. As already discussed with @szarnekow above, the fix should also be applied on the content assistant lexer.

**Comment by tomaswolf** [comment]
The loop can also be avoided with sam974's grammar if you give the lexer a catch-all fallback terminal:
```
terminal ANY_OTHER  : .;
```
Such a fallback terminal exists in `org.eclipse.xtext.common.Terminals`, which explains why this problem is rarely hit.

**Comment by miklossy** [comment]
As @szarnekow suggested in his [comment](https://github.com/eclipse/xtext-eclipse/issues/678#issuecomment-390233984) I created the following two Pull Requests:

1. The PR in `xtext-core` provides a fix to avoid infinite loop in case of an `org.antlr.runtime.FailedPredicateException`. The PR basically provides the same [fix](https://github.com/eclipse/xtext-core/commit/34d61d4d80c6992045ed013a0ac4eb337a1f4e87) already applied for the runtime lexer also for the content assistant lexer.
2. The PR in `xtext-eclipse` provides the [xmleditor](https://github.com/itemis/xtext-issues/tree/issues/xtext-eclipse/690) test language with corresponding test cases. The xmleditor language is based on a stateful lexer demonstrating the occurrence of the `org.antlr.runtime.FailedPredicateException`.

**Comment by miklossy** [comment]
The infinite loop caused by the `FailedPredicateExcpetion` has been fixed.

However, the infinite loop caused by the `EarlyExitException` is still there. It should be investigated if the problem could be solved by a warning on a missing `terminal ANY_OTHER  : .;` as [suggested](https://github.com/eclipse/xtext-eclipse/issues/678#issuecomment-393355235) by @tomaswolf 

See also the [comment](https://github.com/eclipse/xtext-core/issues/69#issuecomment-413887136) from @ArneDeutsch on a similar [issue](https://github.com/eclipse/xtext-core/issues/69)

**Comment by cdietrich** [comment]
will clear milestone
